File: NextPermutation.py
# encoding: utf-8
"""
Implement next permutation, which rearranges numbers into the lexicographically next greater permutation of numbers.

If such arrangement is not possible, it must rearrange it as the lowest possible order (ie, sorted in ascending order).

The replacement must be in-place, do not allocate extra memory.

Here are some examples. Inputs are in the left-hand column and its corresponding outputs are in the right-hand column.
1,2,3 → 1,3,2
3,2,1 → 1,2,3
1,1,5 → 1,5,1

0) Initialize partition_idx = change_idx = -1
1) From right to left, find the first number that violate the ascending trend, called partition_idx
2) if partition_idx exists, from right to left, find the first number greater than nums[partition_idx],
   called change_idx
3) If change_idx, then swap nums[partition_idx], nums[change_idx]
4) Reverse the list on the right of partition_idx (does not include partition_idx)
"""


class Solution:
    # @param {integer[]} nums
    # @return {void} Do not return anything, modify nums in-place instead.
    def nextPermutation(self, nums):
        partition_idx = change_idx = -1
        for idx in xrange(len(nums) - 2, -1, -1):
            if nums[idx] < nums[idx+1]:
                partition_idx = idx
                break

        if partition_idx != -1:
            for idx in xrange(len(nums) - 1, -1, -1):
                if nums[idx] > nums[partition_idx]:
                    change_idx = idx
                    break

        if -1 not in (partition_idx, change_idx):
            nums[partition_idx], nums[change_idx] = nums[change_idx], nums[partition_idx]

        i, j = partition_idx + 1, len(nums) - 1
        while i < j:
            nums[i], nums[j] = nums[j], nums[i]
            i += 1
            j -= 1


if __name__ == '__main__':
    nums = [1, 2, 3]
    Solution().nextPermutation(nums)
    print nums

File: Leetcode-Java/NextPermutation.java
/**
 * Created by yhf on 15/5/23.
 */

/*
Implement next permutation, which rearranges numbers into the lexicographically next greater permutation of numbers.

If such arrangement is not possible, it must rearrange it as the lowest possible order (ie, sorted in ascending order).

The replacement must be in-place, do not allocate extra memory.

Here are some examples. Inputs are in the left-hand column and its corresponding outputs are in the right-hand column.
1,2,3 → 1,3,2
3,2,1 → 1,2,3
1,1,5 → 1,5,1

http://fisherlei.blogspot.com/2012/12/leetcode-next-permutation.html

6 8 7 4 3 2
1. from right to left, find the first element which violates the increasing trend, called PartitionNumber, that is 6.
2. from right to left, find the first element which is larger than PartitionNumber, called ChangeNumber, that is 7.
3. swap PartitionNumber and ChangeNumber, now becomes 7 8 6 4 3 2
4. reverse the current array on the right of the partition index, now becomes 7 2 3 4 6 8
*/



public class NextPermutation {
    public static void nextPermutation(int[] nums) {
        int partition = -1, change = -1;
        for (int i = nums.length - 2; i >= 0; i--) {
            if (nums[i] < nums[i + 1]) {
                partition = i;
                break;
            }
        }

        if (partition != -1) {
            for (int i = nums.length - 1; i >= 0; i--) {
                if (nums[i] > nums[partition]) {
                    change = i;
                    break;
                }
            }
        }

        if (partition != -1 && change != -1) {
            int temp = nums[partition];
            nums[partition] = nums[change];
            nums[change] = temp;
        }


        for (int i = partition + 1, j = nums.length - 1; i <= j; i++, j--) {
            int temp = nums[i];
            nums[i] = nums[j];
            nums[j] = temp;
        }
    }

    public static void main(String[] args) {
        int[] nums = new int[]{3, 2, 1};
        nextPermutation(nums);
    }
}
